File: src/test/java/samples/geo/CitiesNingCallbackTest.java
package samples.geo;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.asynchttpclient.AsyncHttpClient;
import org.asynchttpclient.ListenableFuture;
import org.asynchttpclient.Response;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import samples.geo.domain.City;

import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.stream.Collectors;

import static org.asynchttpclient.Dsl.asyncHttpClient;


@Disabled
public class CitiesNingCallbackTest {

    private static final Logger LOGGER = LoggerFactory.getLogger(CitiesNingCallbackTest.class);

    private ObjectMapper objectMapper = new ObjectMapper();
    private AsyncHttpClient asyncHttpClient = asyncHttpClient();
    private Executor executor = Executors.newWorkStealingPool();

    @Test
    public void callbackHellTest() throws Exception {
        ListenableFuture<Response> responseListenableFuture = asyncHttpClient
                .prepareGet("http://localhost:9090/cityids")
                .execute();

        responseListenableFuture.addListener(() -> {
            try {
                Response response = responseListenableFuture.get();
                String responseBody = response.getResponseBody();
                List<Long> cityIds = objectMapper.readValue(responseBody, new TypeReference<List<Long>>() {});

                cityIds.stream().map(cityId -> {
                    ListenableFuture<Response> cityListenableFuture =
                            asyncHttpClient
                                    .prepareGet("http://localhost:9090/cities/" + cityId)
                                    .execute();

                    return cityListenableFuture.addListener(() -> {
                        try {
                            Response cityDescResp = cityListenableFuture.get();
                            String cityDesc = cityDescResp.getResponseBody();
                            City city = objectMapper.readValue(cityDesc, City.class);
                            LOGGER.info("Got city: " + city);
                        } catch (Exception e) {
                            throw new RuntimeException(e);
                        }
                    }, executor);
                }).collect(Collectors.toList());
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }, executor);


        Thread.sleep(3000);
    }
}

File: src/test/kotlin/samples/geo/kotlin/CitiesNingCallbackTest.kt
package samples.geo.kotlin

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import org.asynchttpclient.Dsl.asyncHttpClient
import org.asynchttpclient.ListenableFuture
import org.asynchttpclient.Response
import org.junit.jupiter.api.Test
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment
import org.springframework.boot.web.server.LocalServerPort
import samples.geo.domain.City
import java.util.concurrent.Executors
import java.util.stream.Collectors


@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
class CitiesNingCallbackTest {

    @Autowired
    private lateinit var objectMapper: ObjectMapper
    private val asyncHttpClient = asyncHttpClient()
    private val executor = Executors.newWorkStealingPool()

    @LocalServerPort
    private val localServerPort: Int? = null

    @Test
    fun callbackHellTest() {
        val responseListenableFuture: ListenableFuture<Response> = asyncHttpClient
                .prepareGet("http://localhost:$localServerPort/cityids")
                .execute()

        responseListenableFuture.addListener(Runnable {
            val response: Response = responseListenableFuture.get()
            val responseBody: String = response.responseBody
            val cityIds: List<Long> = objectMapper.readValue<List<Long>>(responseBody,
                    object : TypeReference<List<Long>>() {})

            cityIds.stream().map { cityId ->
                val cityListenableFuture = asyncHttpClient
                        .prepareGet("http://localhost:$localServerPort/cities/$cityId")
                        .execute()

                cityListenableFuture.addListener(Runnable {
                    val cityDescResp = cityListenableFuture.get()
                    val cityDesc = cityDescResp.responseBody
                    val city = objectMapper.readValue(cityDesc, City::class.java)
                    LOGGER.info("Got city: $city")
                }, executor)
            }.collect(Collectors.toList())
        }, executor)

        Thread.sleep(3000)
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(CitiesNingCallbackTest::class.java)
    }
}

